### PR TITLE
Unions (Horde and Alliance)

### PR DESCRIPTION
**Pull request by zumbak04**
## Changelog:
- Added Unions: the Horde and Alliance. Each member of the union is allied to other members. It features:
  - Joining
    - The AI joins depending on its faith, culture, opinion, and ongoing invasion.
    - The AI doesn't join if the union has more than 6 members. During an invasion, more than 10.
  - Leaving
    - The AI leaves depending on its opinion.
  - Inviting
    - The AI leader doesn't invite. Only players can do this.
    - The AI response depends on its opinion, culture, religion, your prestige, and many more. The most of triggers are taken from the vanilla Negotiate Alliance interaction.
  - Inheriting
    - Both member and leader status is inherited.
  - The historical Horde and Alliance exist in the 603 and 605 startdates.
- Added a new sidebar tab for the Horde and Alliance, the Unions. It shows the current members, leader, regent, heir, and military strength. It can be used to join, leave, and found unions.
- Nerfed the Horde strength in the First War a bit so the Alliance has chances to win.

# Tasklist:
In short, I port functions from CK2.
- [x] Basic effects and triggers for joining, leaving, becoming a leader
- [x] Limiting joining if you're in Pandaria and Uldum
- [x] Founding the Alliance
- [x] Joining the Alliance
- [x] Leaving the Alliance
- [x] Founding the Horde
- [x] Joining the Horde
- [x] Leaving the Horde
- [x] Notification events when the Horde and Alliance founded (texts may come from CK2)
- [x] Notification on becoming the leader
- [x] Notification on destruction
- [x] Faction interface
- [x] Faction history: both factions exist in the year 603 and 605
- [x] Add Political Faction as a game concept (`common\game_concepts`)
- [x] Balance Horde Invasion troops
- [x] Add Invite button
- [x] Wildhammer join the Horde
- [x] Decide who becomes the next faction leader
- [x] Adjust this message
- [x] Block alliances breaking
- [x] Add union interaction cooldown
- [x] Fix errors after leader death
<details>
<summary>Click to expand</summary>

image

</details>

# How to test:
The AI should form unions when Stormwind falls.

Test all the features. Any kind of feedback is welcome: design, balance, lore, bug reports.

[Filter](https://github.com/Warcraft-GoA-Development-Team/Warcraft-Guardians-of-Azeroth-2/pull/116/files?file-filters%5B%5D=.yml) to hide non-localization files for @Warcraft-GoA-Development-Team/english-proofreaders.

### CODE REVIEW

File: localization/english/wc_pol_faction_l_english.yml [hunk 1]
```diff
@@ -0,0 +1,82 @@
+﻿l_english:
+ found_alliance_political_faction:0 "Forge the Alliance"
+ found_alliance_political_faction_desc:0 "We're in danger. If we are to weather the storm, we must make the strongest alliances with our neighbours."
```
Comment (LordSilvermane) [inline localization/english/wc_pol_faction_l_english.yml:3]: Minor change
```suggestion
 found_alliance_political_faction_desc:0 "We're in danger. If we are to weather the storm, we must found new, strong alliances with our neighbours."
```

File: localization/english/wc_pol_faction_l_english.yml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ leave_horde:0 "Leave the Horde"
+ leave_horde_desc:0 "$leave_alliance_desc$"
+ leave_horde_tooltip:0 "$leave_alliance_tooltip$"
+ leave_horde_confirm:0 "$leave_alliance_confirm$"
+ 
+ THE_ALLIANCE:0 "the Alliance"
+ THE_HORDE:0 "the Horde"
+ 
+ WCPLF_100_TITLE:0 "[faction_leader.Custom('PolFaction')|U]"
+ WCPLF_100_DESC_beg:0 "Your messenger brings news that [faction_leader.GetFullNameNicknamed] announces that a new world power raises.\n\n"
```
Comment (LordSilvermane) [inline localization/english/wc_pol_faction_l_english.yml:39]: Minor change
```suggestion
 WCPLF_100_DESC_beg:0 "Your messenger brings news that [faction_leader.GetFullNameNicknamed] announces that a new world power is rising.\n\n"
```

File: localization/english/wc_pol_faction_l_english.yml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+ WCPLF_100_DESC_alliance:0 "They call themselves the Alliance and welcome noble nations of Azeroth who are to weather the coming storm and preserve their cultures and peace. Driven by tradition, they respect deep commitments to abstract concepts like nobility, honor, faith, justice, and sacrifice. Striving to represent these high ideals, they contribute their technical, arcane, and spiritual wisdom toward the shared goal of preserving order on Azeroth and creating a peaceful and just world."
```
Comment (LordSilvermane) [inline localization/english/wc_pol_faction_l_english.yml:40]: Minor changes
```suggestion
 WCPLF_100_DESC_alliance:0 "They call themselves the Alliance and welcome all noble nations of Azeroth who are struggling to weather the storm and preserve their cultures and peace. Driven by tradition, they respect deep commitments to abstract concepts like nobility, honor, faith, justice, and sacrifice. Striving to represent these high ideals, they declare their willingness to contribute their technical, arcane, and spiritual wisdom toward the shared goal of preserving order on Azeroth and creating a peaceful and just world."
```

File: localization/english/wc_pol_faction_l_english.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+ WCPLF_100_DESC_horde:0 "They call themselves the Horde and welcome disparate nations that struggle to survive in the hostile world that would see them destroyed. Driven by unity, they are bonded together as a family, comrades, or even uneasy allies. Focused, ferocious, and sometimes monstrous, the Horde values strength and honor and is ready to push back the ones threatening their ideals and shatter the status quo."
```
Comment (LordSilvermane) [inline localization/english/wc_pol_faction_l_english.yml:41]: Minor changes
```suggestion
 WCPLF_100_DESC_horde:0 "They call themselves the Horde and welcome all disparate nations that are struggling to survive in this hostile world that would see them destroyed. Driven by unity, they are bonded together as a family, comrades, or even uneasy allies. Focused, ferocious, and sometimes monstrous, the Horde values strength and honor and is ready to push back those that threaten their ideals and seek to shatter the status quo."
```

File: localization/english/wc_pol_faction_l_english.yml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+ WCPLF_100_OPT_neg:0 "[faction_leader.Custom('PolFaction')|U] you say... another enemy to fight with."
+ WCPLF_100_OPT_neu:0 "[faction_leader.Custom('PolFaction')|U] you say... keep an eye on them."
+ FOR_FACTION:0 "For [faction_leader.Custom('PolFaction')]!"
+ 
+ WCPLF_150_DESC:0 "My nation and I would like to join your side and become a part of [ROOT.Char.Custom('PolFaction')]."
+ 
+ WCPLF_155_DESC:0 "[recipient.Custom('PolFaction')|U] welcomes you! It would be an honor to fight side by side with you."
+ WCPLF_160_DESC:0 "After long consideration, I am afraid we have come to the conclusion that the [recipient.Custom('PolFaction')|U] must refuse your offer."
+ 
+ WCPLF_200_DESC:0 "It feels like [faction_leader.Custom('PolFaction')] has lost its leaders.\n\nAmong all who took the oath, your nation is the most respectful and strong. Your allies expect you to lead [faction_leader.Custom('PolFaction')] before it falls apart."
```
Comment (LordSilvermane) [inline localization/english/wc_pol_faction_l_english.yml:52]: Minor changes
```suggestion
 WCPLF_200_DESC:0 "It feels like [faction_leader.Custom('PolFaction')] has lost its leaders.\n\nAmong all who took the oath, your nation is the most respected and the strongest. Your allies expect you to lead [faction_leader.Custom('PolFaction')] before it falls apart."
```

File: localization/english/wc_pol_faction_l_english.yml [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+ 
+ WCPLF_250_DESC:0 "[me.Custom('PolFaction')|U] has lost its leaders.\n\nFinding itself in a state of chaos, unable to find anyone to wear this burden, [me.Custom('PolFaction')] collapsed. The former members have stated that their previous agreements are no longer valid."
```
Comment (LordSilvermane) [inline localization/english/wc_pol_faction_l_english.yml:54]: Slight change - not totally necessary but makes more sense
```suggestion
 WCPLF_250_DESC:0 "[me.Custom('PolFaction')|U] has lost its leaders.\n\nFinding itself in a state of chaos, unable to find anyone to shoulder this burden, [me.Custom('PolFaction')] collapsed. The former members have stated that their previous agreements are no longer valid."
```

File: localization/english/wc_pol_faction_l_english.yml [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ I_HAVE_NO_CHOICE:0 "It seems I have little choice..."
+ WCPLF_305_DESC:0 "I am honored by your request, and I would be glad to join the [actor.Custom('PolFaction')]."
+ MARVELOUS_NEWS:0 "Marvelous news!"
+ WCPLF_310_DESC:0 "After long consideration, I am afraid I have come to the conclusion that I must refuse your offer."
+ TROUBLING:0 "Troubling..."
+ 
+ invite_to_pol_faction_interaction:0 "Invite to Union"
+ invite_to_pol_faction_interaction_extra_icon:0 "$hook_extra_icon$"
+ invite_to_pol_faction_interaction_desc:0 "Invite [recipient.GetShortUIName] to your current [union|E], [actor.Custom('PolFaction')]"
+ invite_to_pol_faction_interaction_notification:0 "I eagerly propose to join my side and become a part of the [actor.Custom('PolFaction')]."
```
Comment (LordSilvermane) [inline localization/english/wc_pol_faction_l_english.yml:75]: Minor change
```suggestion
 invite_to_pol_faction_interaction_notification:0 "I eagerly propose that you join my side and become a part of the [actor.Custom('PolFaction')]."
```

**Review by LordSilvermane — COMMENTED** [review]
A few suggested fixes and changes to localization

**Review by LordSilvermane — APPROVED** [review]
Loc looks good! LGTM!

**Review by TheDrPibb1999 — APPROVED** [review]
My boy Silvermane seemed to make some good changes to the loc, I read it and saw no main issues :) lgtm

**Review by TheGroxEmpire — COMMENTED** [review]
**Bugs**
- The Horde Union is missing of Regent while the invasion is happening:
<details>

image
</details>

- The loc for joining the Horde is not showing properly:
<details>

image
</details>

- If you die while the Unions tab are opened, it'll appear blank before you close and reopen the tab again. Paradox solved this issue by having other tabs closed after we die:
<details>

image
</details>

**Suggestion**
- I think leaving the union should give some opinion penalties to the union members and its leader.
- Garithos was the leader of the Alliance during the Frozen Throne start. Him being the highest ranking left on Lordaeron (https://wow.gamepedia.com/Alliance_resistance).
- I think Unions member should only be called on defensive war / when the union leader declare a great war (Similar to GHW but for union. Something we can implement later maybe?). The AI are trigger happy with the call to war button and subject every union members to their petty offensive war. Stormwind shouldn't call the entire Alliance forces just to apprehend some Gnolls in Redrige for example.

**Comment by zumbak04** [comment]
> * The Horde Union is missing of Regent while the invasion is happening:

It's not a bug. No one can inherit the Horde except for the heir.

**Comment by zumbak04** [comment]
> * I think Unions member should only be called on defensive war / when the union leader declare a great war (Similar to GHW but for union. Something we can implement later maybe?). The AI are trigger happy with the call to war button and subject every union members to their petty offensive war. Stormwind shouldn't call the entire Alliance forces just to apprehend some Gnolls in Redrige for example.

From what I've seen in tests, the AI doesn't invite all. Only some part the AI considers enough to win the war.

**Comment by zumbak04** [comment]
> * The loc for joining the Horde is not showing properly:

Fixed in 7accaea085bf021a99ed7aee59310532f7c02b90.
 
> * If you die while the Unions tab are opened, it'll appear blank before you close and reopen the tab again. Paradox solved this issue by having other tabs closed after we die:

> * I think leaving the union should give some opinion penalties to the union members and its leader.

Changed in 4ecdf7cbe763f2457db7815119a2180ecd8a6e84.

> * Garithos was the leader of the Alliance during the Frozen Throne start. Him being the highest ranking left on Lordaeron (https://wow.gamepedia.com/Alliance_resistance).

Changed in a851ebee3f50f64545c468384815c00b68d79aa6.

**Review by TheGroxEmpire — COMMENTED** [review]
- Should probably removes the effect where you lose opinion of yourself when leaving union. Or at least remove the text of it.
<details>

image
</details>

- Maybe we should put something here in the Effects box:
<details>

image
</details>

Aside from that other bugs seem to be fixed and the rest seems fine.

File: localization/english/wc_game_concepts_l_english.yml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+﻿l_english:
+ game_concept_union:0 "Union"
+ game_concept_unions:0 "Unions"
+ game_concept_union_possessive:0 "Union's"
```
Comment (arithon) [inline localization/english/wc_game_concepts_l_english.yml:4]: could we go without the possessive and use "of the Union" instead? Would make german (probably french and other languages too) more easy to translate...

File: localization/english/wc_gui_l_english.yml [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿l_english:
+ CV_HAS_NO_DIPLOMACY:0 "Has no ongoing diplomacy"
+ CV_ALLIANCE:0 "Alliance"
+ CV_HORDE:0 "Horde"
+ CV_MEMBERS:0 "Members #weak ([GetPlayer.MakeScope.Var('gui_pol_faction_member_number').GetValue|V|0])#!"
+ CV_LEADER:0 "[leader_wo_union|E]"
+ CV_UNION_SOLDIERS:0 "Military Strength: [GetPlayer.MakeScope.Var('gui_pol_faction_strength').GetValue|V|0]"
+ 
+ POL_FACTIONS_WINDOW_TITLE:0 "Unions"
+ HUD_POL_FACTIONS:0 "Unions"
```
Comment (arithon) [inline localization/english/wc_gui_l_english.yml:10]: Why not use " [union|E] " here instead of "Union", just in case something changes eventually?

**Comment by arithon** [comment]
In the union tab, I can't click on the heir, is that just a bug on my end or do others experience the same thing?

**Comment by zumbak04** [comment]
> * Should probably removes the effect where you lose opinion of yourself when leaving union. Or at least remove the text of it.

Fixed in e1de8c4ac58fb564e5d8f6c61043d7cb41dae17f.

> * Maybe we should put something here in the Effects box:

Changed in e1de8c4ac58fb564e5d8f6c61043d7cb41dae17f.

> * In the union tab, I can't click on the heir, is that just a bug on my end or do others experience the same thing?
> * When hovering over empty space on the union tab, I get the tooltip of the land behind the tab.

Fixed in b2439e1ad6b03a100d71ceb89e4130f87a75bd30.

File: localization/english/wc_gui_l_english.yml (hunk 9 above)
Comment (zumbak04) [reply localization/english/wc_gui_l_english.yml]: Changed in 74299d9bf6e6decbc1a602c43b19b8242bcad869.

**Comment by arithon** [comment]
Why does the cult_of_sunwell faith not prefer the alliance union?

**Review by arithon — APPROVED** [review]
loc LGTM, leaves only the question why elves prefer another union then the alliance

**Comment by zumbak04** [comment]
> loc LGTM, leaves only the question why elves prefer another union then the alliance

So they may join the Horde later.

**Review by arithon — APPROVED** [review]
LGTM

**Review by TheGroxEmpire — COMMENTED** [review]
- The effect text is missing a point vanilla:
<details>

image

Vanilla example:

image
image
</details>

That should be the last from me.

**Review by TheGroxEmpire — APPROVED** [review]

**Pull request merged**